docs/plutus-ir/plutus-ir.md

```diff
@@ -0,0 +1,193 @@
+list only the additions and removals, marking them explicitly with `+` or `-`.
+We also require syntax with lists of subparts. We denote a list of `t`-productions with `t..`.
+
+```
+VarDecl VD      := (vardecl x T)
+TyVarDecl TVD   := (tyvardecl alpha K)
+
+Type R, S, T    := - (fix alpha T)
+Term L, M, N    := + (let BK B.. M)
+
+BiningKind BK   := (nonrec)
+                   (rec)
+Binding B       := (termbind VD L)
+                   (typebind TVD T)
+                   (datatypebind D)
+
+Datatype D      := (datatype
+                      TVD   # the type variable for the datatype
+                      TVD.. # the argument type variables of the datatype
+                      x     # the name of the destructor function
+                      C..   # the constructors of the datatype
+                   )
+Constructor C   := (constr
+                       x   # the name of the constructor
+                       T.. # the argument types of the constructor
+                    )
+```
+
+Here is an example, constructing `[1]`:
+```
+# bind List
+(let (rec)
+    (datatypebind
+        (datatype
+            (tyvardecl List (fun (type) (type)))
+            (tyvardecl a (type))
+            match_List
+            (constr
+                Nil
+            )
+            (constr
+                Cons
+                a
+                [List a]
+            )
+        )
+    )
+    # bind int64 for convenience
+    (typebind (tyvardecl int (type)) [(con integer) (con 64)])
+    # apply cons to 1 and nil
+    [ { Cons int } (con 64 ! 1) { Nil int } ]
+)
+```
+
+Another, implementing `fromMaybe`:
+```
+# bind Maybe
+(let (nonrec)
+    (datatypebind
+        (datatype
+            (tyvardecl Maybe (fun (type) (type)))
+            (tyvardecl a (type))
+            match_Maybe
+            (constr
+                Nothing
+            )
+            (constr
+                Just
+                a
+            )
+        )
+    )
+    (abs a (type)
+        (lam default a
+            (lam arg [Maybe a]
+                # match on the arg, producing an a, giving the default
+                # in the Nothing case and using the identity function
+                # in the Just case
+                [ { match_Maybe a } arg default (lam x a x) ]
+            )
+        )
+    )
+)
+```
+
+
+## Validity
+
+### Mutually recursive types
+
+The syntax of Plutus IR suggests that you can write mutually recursive datatypes, even though
+we don't know how to compile those yet. I think we will eventually be able to cover all
+types defined with this syntax, but until then, validity of Plutus IR programs will be somewhat
+determined simply by *whatever we are able to compile*.
+
+### Typechecking
+
+Similarly, while we will define a typing judgement for Plutus IR, we will probably not
+write a typechecker for it, and rather will compile it to Plutus Core before typechecking.
```

Comment:
Regarding Value Restriction, below. As I recall, we have two versions of Plutus Core. One does not have the value restriction, and can be used to normalise all types, yielding code that does satisfy the value restriction. So if Plutus IR targets the non-normalised version there is no problem, and we can still generate the normalised version. Is that correct?

Reply:
By the "value restriction" I mean the restriction that type abstractions can only have term values as their bodies. I don't think this relates to the type normalization stuff?